### PR TITLE
Smart contracts phase3

### PR DESCRIPTION
**Pull request by Krestol**
1. Super Agents
2. Functions for saving information about successful operations
3. Function for saving adderess for rewards

### CODE REVIEW

File: contracts/OnyxPay.py [hunk 1]
```diff
@@ -1,33 +1,160 @@
...
+        rewardBalance = args[1]
+        return RegisterRewardBalance(agentId, rewardBalance)
+    
+    if operation == "GetRewardBalance":
+        if len(args) != 1:
+            return False
+        agentId = args[0]
+        return GetRewardBalance(agentId)
+
+    if operation =="SuccessfulOperation":
```
Comment (arsonaut) [inline contracts/OnyxPay.py]:
```suggestion
    if operation == "SuccessfulOperation":
```

File: contracts/OnyxPay.py [hunk 2]
```diff
@@ -1,33 +1,160 @@
...
+        userId = args[0]
+        return SuccessfulOperation(userId)
+
+    if operation == "GetSuccessfulOperationsCount":
+        if len(args) != 1:
+            return False
+        userId = args[0]
+        return GetSuccessfulOperationsCount(userId)
+
+    if operation =="FailedOperation":
```
Comment (arsonaut) [inline contracts/OnyxPay.py]:
```suggestion
    if operation == "FailedOperation":
```

File: contracts/OnyxPay.py [hunk 3]
```diff
@@ -1,33 +1,160 @@
...
+    return True if Get(context, superagentId) == superAgentMagic else False
+
+
+def RegisterRewardBalance(agentId, rewardBalance):
+    WitnessRequire(agentId)
+    Put(context, ConcatKey(agentId, rewardBalanceMagic), rewardBalance) 
+
+
+def GetRewardBalance(agentId):
+    rewardBalance = Get(context, ConcatKey(agentId, rewardBalanceMagic))
```
Comment (arsonaut) [inline contracts/OnyxPay.py]: It is more safe to use key prefixes, and use prefixes for all storage.

File: tests/test_OnyxPay.py [hunk 4]
```diff
@@ -1,43 +1,259 @@
...
+        OnyxPayWrapper.successfulOperation(WalletWrapper.AliceAddress(), WalletWrapper.Admin())
+        SdkUtils.WaitNextBlock()
+        self.assertEqual(OnyxPayWrapper.getSuccessfulOperationsCount(WalletWrapper.AliceAddress()), oldSuccessfulOperations + 1)
+
+
+    def test_several_successful_operations(self):
+        oldSuccessfulOperations = OnyxPayWrapper.getSuccessfulOperationsCount(WalletWrapper.AliceAddress())
+        successfulOperations = 2
+        i = 0
+        while i < successfulOperations:
```
Comment (arsonaut) [inline tests/test_OnyxPay.py]:
```suggestion
        for i in range(successfulOperations):
```

File: contracts/OnyxPay.py (hunk 3 above)
Comment (Krestol) [reply contracts/OnyxPay.py]: could you explain in which way prefix form is more safety then postfix?

Comment (arsonaut) [reply contracts/OnyxPay.py]: Ok, I'm correcting myself, it is unsafe if:
* postfixes (or suffixes) are not used in all cases, only in some
* userId's can be arbitrary in size

Then malicious user can craft userId witch will end with suffix (or start with postfix) and such userId will make an collision with another key.

In current case it's not relevant due to fixed size of userId's (they are 20 byte addresses).

Comment (Krestol) [reply contracts/OnyxPay.py]: thanks, I will add a check of the length of the address, which is used as user_id

**Pull request merged**